### PR TITLE
refactor based on feedback

### PR DESCRIPTION
**Pull request by michaelb990**
@urmilan & millstone helped with scala, scalatra and slick design feedback. this should provide a cleaner seperation between the data access and business logic/service layers.

### CODE REVIEW

File: src/main/scala/com/socrata/decima/data_access/DeployAccess.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package com.socrata.decima.data_access
+
+import com.socrata.decima.database.{DatabaseDriver, DeployDAO}
+import com.socrata.decima.models._
+import slick.driver.PostgresDriver.simple.Database
+
+
+trait DeployAccess {
+  def createDeploy(deploy:DeployForCreate):Either[Exception, Deploy]
```
Comment (urmilan) [inline src/main/scala/com/socrata/decima/data_access/DeployAccess.scala]: Style nit - the convention in Scala is to have a space between : and type eg. 
`def createDeploy(deploy: DeployForCreate): Either[Exception, Deploy]`

File: src/main/scala/com/socrata/decima/data_access/DeployAccess.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  def currentDeploymentState(environment:Option[String], service:Option[String]):Seq[Deploy]
+  def deploymentHistory(environment:Option[String], service:Option[String], limit: Int):Seq[Deploy]
+}
+
+case class DeployAccessWithPostgres(db:Database, dao:DeployDAO with DatabaseDriver) extends DeployAccess {
+  override def createDeploy(deploy: DeployForCreate): Either[Exception, Deploy] = {
```
Comment (urmilan) [inline src/main/scala/com/socrata/decima/data_access/DeployAccess.scala]: the braces surrounding the method are redundant.

File: src/main/scala/com/socrata/decima/database/DeployDAO.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      val newId = (deployTable returning deployTable.map(_.id)) += DeployRow(0,
+                                                                              deploy.service,
+                                                                              deploy.environment,
+                                                                              deploy.version,
+                                                                              deploy.revision,
+                                                                              deploy.deployedBy,
+                                                                              deploy.deployMethod,
+                                                                              TimeUtils.asTimestamp(now))
+      val newDeploy = DeployCompiledQueries.lookup(newId).run.headOption
+      newDeploy match {
```
Comment (urmilan) [inline src/main/scala/com/socrata/decima/database/DeployDAO.scala]:
```scala
newDeploy.map(Right(rowToModelDeploy)).getOrElse(Left(...))
```

File: src/main/scala/com/socrata/decima/database/DeployDAO.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        case Some(d) => Right(rowToModelDeploy(d))
+        case None => Left(new RuntimeException("Unable to create deploy"))
+      }
+    }
+  }
+
+  def lookup(id: Long)(implicit session:Session): Either[Exception, Deploy] = {
+    val deploy = DeployCompiledQueries.lookup(id).run.headOption
+    deploy match {
```
Comment (urmilan) [inline src/main/scala/com/socrata/decima/database/DeployDAO.scala]: ditto ^^^

File: src/main/scala/com/socrata/decima/database/DeployDAO.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+  def currentDeployment(environment: Option[String], service: Option[String])(implicit session:Session): Seq[Deploy] = {
+    val res = currentDeploymentQuery.list.filter(row => environment match {
+      case Some(e) => row.environment == e
+      case None => true
+    }).filter(row => service match {
+      case Some(s) => row.service == s
+      case None => true
+    })
+    res.map(d => rowToModelDeploy(d))
```
Comment (urmilan) [inline src/main/scala/com/socrata/decima/database/DeployDAO.scala]:
```scala
res.map(rowToModelDeploy)
```

File: src/main/scala/com/socrata/decima/services/DeployService.scala [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+   * A PUT call to /deploy takes a JSON object with the following keys:
+   * "service": name of service being deployed
+   * "environment": the environment being deployed to
+   * "version": version being deployed
+   * "git": the git SHA of the deploy
+   */
+  put("/") {
+    val deploy = parsedBody.extract[DeployForCreate]
+
+    deployAccess.createDeploy(deploy)
```
Comment (urmilan) [inline src/main/scala/com/socrata/decima/services/DeployService.scala]: Is there something useful you could return in the response here, like an ID?

File: src/main/scala/com/socrata/decima/util/TimeUtils.scala [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import org.joda.time.{DateTime, DateTimeZone}
+
+object TimeUtils {
+
+  def toJodaTime(timestamp:Timestamp): DateTime = {
+    val timeZone = DateTimeZone.UTC
+    new DateTime(timestamp.getTime, timeZone)
+  }
+
+  def asTimestamp(jodaTime: DateTime): Timestamp = {
```
Comment (urmilan) [inline src/main/scala/com/socrata/decima/util/TimeUtils.scala]: I would call this `asSqlTimestamp`, just so it's clearer why you're calling this.

File: src/test/scala/com/socrata/decima/DeployDAOSpec.scala [hunk 8]
```diff
@@ -0,0 +1,147 @@
+package com.socrata.decima
+
+import com.socrata.decima.models.{DeployForCreate, Deploy}
+import org.slf4j.LoggerFactory
+
+import scala.slick.driver.H2Driver
```
Comment (urmilan) [inline src/test/scala/com/socrata/decima/DeployDAOSpec.scala]: Nice :+1:

File: src/main/scala/com/socrata/decima/services/DeployService.scala (hunk 6 above)
Comment (michaelb990) [inline src/main/scala/com/socrata/decima/services/DeployService.scala]: I return the full object (w/ id). I'll add a comment.

File: src/main/scala/com/socrata/decima/database/DeployDAO.scala (hunk 3 above)
Comment (michaelb990) [inline src/main/scala/com/socrata/decima/database/DeployDAO.scala]: for some reason that syntax doesn't seem to be working... I can do:
`deploy.map(d => Right(rowToModelDeploy(d))).getOrElse(Left(new RuntimeException(...`
but not
`deploy.map(Right(rowToModelDeploy)).getOrElse(Left(new RuntimeException(...`

What's the secret!?

Comment (urmilan) [inline src/main/scala/com/socrata/decima/database/DeployDAO.scala]: Oh, my bad. I guess Scala cannot figure out that we want pass the parameter in once it is wrapped  in the `Right`.

File: src/main/scala/com/socrata/decima/services/DeployService.scala (hunk 6 above)
Comment (urmilan) [inline src/main/scala/com/socrata/decima/services/DeployService.scala]: Oh, right, I did't check back to see if createDeploy returns something :+1:

**Comment by urmilan** [comment]
:shipit:

**Pull request merged**